<div style="border:solid green 2px; padding: 20px">
<b>Hello, Pranita!</b>
    
Grigoriy is here, I've reviewed щту of your previous projects. There were so many submissions to review at the final stage that the regular reviewers were asked to get involved in this process. Don't worry, your tutor will be reviewing the major project anyway.  
    
You're all data analysts now, so no yellow comments here. In my feedback I tried to give the comments in the way your future team lead would do.    

You can find my comments in <font color='green'>green</font> or <font color='red'>red</font> boxes like this:

<div class="alert alert-success">
<b>Reviewer's comment:</b> if everything is done succesfully
</div>

<div class="alert alert-danger">
<b>Reviewer's comment:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>
    
Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:
    
<div class="alert alert-info"> <b>Student's comment:</b> For example like this.</div>    
    
      
<div>

### <font color='orange'>General feedback</font>
Thank you for sending your project! Well done, all queries are correct, but pay attention that you accidentally skipped one of the questions. I've marked the problem point below.

**Update**:<br>
Thank you for the update! Your project has been accepted, congratulations!

---

# SQL Project

### Project description:

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.

We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.


### Data description:
There were five datasets provided.

##### books contains
- book_id 
- author_id
- title
- num_pages 
- publication_date
- publisher_id

##### authors contains
- author_id
- author 

##### publishers contains
- publisher_id
- publisher 

##### ratings contains
- rating_id
- book_id 
- username 
- rating

##### reviews contains
- review_id
- book_id 
- username 
- text

#### Task

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating: look only at books with at least 50 ratings.
- Find the average number of text reviews among users who rated more than 50 books.

#### Goal:
Our goal is to make decisions for startups to develop new apps for book lovers by providing the details of books world.


<div class="alert alert-success">
<b>Reviewer's comment: </b> Nice introduction!
</div>


### Importing Libraries:

In [38]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.dates
import seaborn as sns
import math
import matplotlib.dates as mdates
import datetime 
import math
from sqlalchemy import create_engine


### Access to the database:

In [39]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Studying  the tables:

In [40]:
books = pd.io.sql.read_sql('''SELECT * FROM books''', con = engine)

display(books.info())
display(books.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

#### Conclusion:
The 'books' table contain 1000 rows of book information. Each row of book has book_id, title,number of pages,publication date and publisher id.


In [41]:
authors = pd.io.sql.read_sql('''SELECT * FROM authors''', con = engine)
display(authors.info())
display(authors.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

#### Conclusion:
The 'authors' table contain 636 rows of author information. Each row of this table has author id and author name.


In [42]:
publishers = pd.io.sql.read_sql('''SELECT * FROM publishers''', con = engine)
display(publishers.info())
display(publishers.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

#### Conclusion:
The 'publishers' table contain 340 rows of publishers information. Each row of this table has publisher id and publishers name.

In [43]:
ratings = pd.io.sql.read_sql('''SELECT * FROM ratings''', con = engine)
display(ratings.info())
display(ratings.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

#### Conclusion:
The 'ratings' table contain 6456 rows of ratings information. Each row of this table has rating id and book id of the book that a user rated, username of the  name and rating score.

In [44]:
reviews = pd.io.sql.read_sql('''SELECT * FROM reviews''', con = engine)
display(reviews.info())
display(reviews.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

#### Conclusion:
The 'reviews' table contain 2793 rows of book review information. Each row of this table has review id and book id of the book that a user rated, username of the  name and review text.

<div class="alert alert-success">
<b>Reviewer's comment: </b> Well done, let's look at the SQL queries.
</div>

### Making SQL Queries for Each Task:

In [45]:
# Find the number of books released after January 1, 2000.

books_released = pd.io.sql.read_sql('''SELECT Count(distinct book_id) FROM books WHERE publication_date > '2000-01-01' ''', con = engine)
display(books_released)

count
0    819

#### Conclusion:
The number of books released after January 1st, 2000 is 819

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

In [46]:
#Find the number of user reviews and the average rating for each book.

query = ''' SELECT books.book_id, books.title, 
avg( distinct ratings.rating) as avg_rating, 
count(distinct reviews.review_id) as user_reviews
FROM  books 
INNER JOIN ratings ON books.book_Id = ratings.book_id
INNER JOIN reviews ON books.book_id = reviews.book_id
GROUP BY books.book_id
ORDER BY  avg_rating DESC'''

result = pd.io.sql.read_sql(query, con = engine)
display(result)

book_id                                              title  avg_rating  \
0        732  The Ghost Map: The Story of London's Most Terr...         5.0   
1        136  Captivating: Unveiling the Mystery of a Woman'...         5.0   
2         17                      A Dirty Job (Grim Reaper  #1)         5.0   
3        330  How to Be a Domestic Goddess: Baking and the A...         5.0   
4         57                    Act of Treason (Mitch Rapp  #9)         5.0   
..       ...                                                ...         ...   
989      164                            Count Zero (Sprawl  #2)         2.5   
990      772                             The Kitchen God's Wife         2.0   
991      371                                              Junky         2.0   
992      316                  His Excellency: George Washington         2.0   
993      303                               Harvesting the Heart         1.5   

     user_reviews  
0               2  
1               2  
2               4  
3               1  
4               2  
..            ...  
989             2  
990             3  
991             2  
992             2  
993             2  

[994 rows x 4 columns]

#### Conclusion:
The above table shows a list of 994 books with the number of user reviews and the average rating for each book. 

---

<div class="alert alert-danger">
<s><b>Reviewer's comment:</b> Please, answer to this question: <br>
Find the number of user reviews and the average rating for each book.
</div>

---

In [47]:
# Identify the publisher that has released the greatest number of books with more than 50 pages.

query = ''' SELECT publishers.publisher, count(distinct books.book_id) as no_of_books
FROM publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50 
GROUP BY publishers.publisher
ORDER BY no_of_books DESC '''

result = pd.io.sql.read_sql(query, con = engine)
print(result)

                               publisher  no_of_books
0                          Penguin Books           42
1                                Vintage           31
2               Grand Central Publishing           25
3                       Penguin Classics           24
4                                 Bantam           19
..                                   ...          ...
329                  Random House Anchor            1
330  Random House Audio Publishing Group            1
331  Random House Large Print Publishing            1
332                   Random House Trade            1
333                 Random House Vintage            1

[334 rows x 2 columns]


<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

#### Conclusion:
The above table shows a list of publishers that have released the greatest number of books with more than 50 pages.


In [48]:
# Identify the author with the highest average book rating: only books with at least 50 ratings.

query = ''' SELECT authors.author as author_name,
                AVG(ratings.rating) as avg_rating
FROM authors
INNER JOIN books ON books.author_id = authors.author_id
INNER JOIN ratings on ratings.book_id = books.book_id
GROUP BY author_name
HAVING COUNT(ratings.rating) > 50
ORDER BY avg_rating DESC '''

result = pd.io.sql.read_sql(query, con = engine)
print(result)

                                          author_name  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.288462
1                                     Agatha Christie    4.283019
2                   Markus Zusak/Cao Xuân Việt Khương    4.264151
3                                      J.R.R. Tolkien    4.240964
4                            Roald Dahl/Quentin Blake    4.209677
5                                   Louisa May Alcott    4.203704
6                                        Rick Riordan    4.130952
7                                       Arthur Golden    4.107143
8                                        Stephen King    4.009434
9                                        John Grisham    3.971429
10                                    William Golding    3.901408
11                                    Nicholas Sparks    3.882883
12                                       Jodi Picoult    3.881579
13                                    Sophie Kinsella    3.877193
14        

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

#### Conclusion:
The above table shows a list of authors with the highest average book rating that is only books with at least 50 ratings.

In [49]:
#Find the average number of text reviews among users who rated more than 50 books.

query = ''' SELECT reviews.username, count( reviews.review_id) as avg_text
FROM reviews
WHERE reviews.username IN (SELECT ratings.username  from ratings GROUP BY ratings.username HAVING COUNT(ratings.rating_id)>50)
GROUP BY reviews.username

ORDER BY avg_text DESC '''

result = pd.io.sql.read_sql(query, con = engine)
print(result)


         username  avg_text
0     sfitzgerald        28
1      martinadam        27
2       richard89        26
3  jennifermiller        25
4          paul88        22
5          xdavis        18


<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

### Conclusion:
 The above table shows a list of users who have the average number of text reviews and rated more than 50 books.

### Conclusion:
 Overall, the queries above provided information such as the number of books released after January 1, 2000, number of user reviews, and the average rating for each book, identify the publisher that has released the greatest number of books with more than 50 pages, identify the author with the highest average book rating and find the average number of text reviews among users who rated more than 50 books. This info helps any startup to generate a value proposition for a new product.